### PR TITLE
[10.0] Module stock_no_negative

### PR DESCRIPTION
**Pull request by alexis-via**
This is in fact a port to v10 of my PR #137 but changing the name to stock_no_negative, because there is already an OCA module with that name in older versions.

### CODE REVIEW

File: stock_no_negative/views/product.xml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+<odoo>
+
+    <record id="view_template_property_form" model="ir.ui.view">
+        <field name="name">stock_no_negative.product.template.form</field>
+        <field name="model">product.template</field>
+        <field name="inherit_id" ref="stock.view_template_property_form" />
+        <field name="arch" type="xml">
+            <field name="categ_id" position="after">
+                <field name="allow_negative_stock"
```
Comment (JordiBForgeFlow) [inline stock_no_negative/views/product.xml:16]: I think that this field should only be accessible to a specific security group. You basically prevent negative stocks, but allow any user to override this rule just going to the product and changing it. What do you think?

**Comment by JordiBForgeFlow** [comment]
We'll add test cases to this PR, and also backport to 9.0

File: stock_no_negative/views/product.xml (hunk 1 above)
Comment (alexis-via) [reply stock_no_negative/views/product.xml:16]: Only few users should have write access on products... as a lot of parameters on products are sensitive, not only this one !

Comment (JordiBForgeFlow) [reply stock_no_negative/views/product.xml:16]: Agreed!

**Review by sebalix — APPROVED** [review]
:+1: 
One note, the `stock_no_negative` module from 8.0 has its field named `check_no_negative` which should be converted to the new `allow_negative_stock`, but this should be done in the v9.0 migration (PR #288).

File: stock_no_negative/models/stock_quant.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# -*- coding: utf-8 -*-
+# © 2015-2016 Akretion (http://www.akretion.com)
+# @author Alexis de Lattre <alexis.delattre@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, api, _
```
Comment (rousseldenis) [inline stock_no_negative/models/stock_quant.py]: Change import order

**Review by rousseldenis — APPROVED** [review]
:+1: Minor comment. Tested on runbot.

File: stock_no_negative/models/stock_quant.py (hunk 2 above)
Comment (alexis-via) [reply stock_no_negative/models/stock_quant.py]: Why should I change import order ?

Comment (rousseldenis) [reply stock_no_negative/models/stock_quant.py]: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#imports

alphabetically ordered. Details but convention ;-)

**Comment by sebalix** [comment]
Is this module ready to be merged?

File: stock_no_negative/models/stock_quant.py (hunk 2 above)
Comment (alexis-via) [reply stock_no_negative/models/stock_quant.py]: Arf !!! Incredible ! Don't count on me to apply these crazy coding conventions !

**Comment by alexis-via** [comment]
@sebalix yes, of course. I've been using it in production for several months.

**Comment by Cedric-Pigeon** [comment]
@pedrobaeza this one can be merged

**Comment by pedrobaeza** [comment]
Maybe you can clean a bit the commits?

**Comment by yvaucher** [comment]
Should this module be tested separately or other module's tests should ensure non zero stock level ?

**Comment by yvaucher** [comment]
See comment https://github.com/OCA/stock-logistics-workflow/pull/341#issuecomment-310319638

**Comment by levkar** [comment]
Works like a charm

**Comment by rousseldenis** [comment]
Setup : https://github.com/akretion/stock-logistics-workflow/pull/3

**Comment by MiquelRForgeFlow** [comment]
We have arrived to the conclusion after some analysis, for the good health of the repository, that the better approach to proceed with this topic is the following: In the _config.settings_ of Stock, you have to select an option to activate the stock_no_negative after the installation of the module. In the Readme is put a configuration note that after the installation, you have to select that option if you want to use the module.

We have developed a patch with this approach for v9, and we will open the PR soon. If you like, you may want to cherry-pick it here.

**Comment by JordiBForgeFlow** [comment]
To add to the point of @mreficent , the tests of module 'stock_no_negative' will activate this setting, and then proceed to do the tests. The change in config.settings should create an entry in ir_config_parameter with name 'stock_no_negative' and value 'True'. 

If the parameter does not exist, or is False, then it is assumed that the negative stocks are allowed, regardless if the products have been set with allow_negative = False.

**Comment by pedrobaeza** [comment]
That's not needed: you just need to deactivate the check if in test mode and you are not explicitly testing this module (which is done through a context key). Check https://github.com/OCA/bank-payment/blob/43f8723f8cb4d8defc50d53e83f8b523a7560f00/account_payment_transfer_reconcile_batch/models/bank_payment_line.py#L28 for an example on how to do it

**Comment by MiquelRForgeFlow** [comment]
@pedrobaeza we will do that way now, thanks!

**Comment by MiquelRForgeFlow** [comment]
Done. @alexis-via please, can you cherry-pick commit 73fa87128b160a4f211800b8404110e76e5b3bfd from https://github.com/OCA/stock-logistics-workflow/pull/357?

**Comment by pedrobaeza** [comment]
Better wait first to have it finished (although of course it works)

**Comment by pedrobaeza** [comment]
Now you can cherry-pick 73fa871

**Comment by levkar** [comment]
Actually disallowing negative stock by default makes more sense in othis module. In v8, we had serious problems with customers because of this behavior and finally we wrote another auto-install module which disallows it by default. 

There was a PR of a new module named stock_disallow_negative but later author decided to use this module name and cancelled that PR. Same rationale.

I would strongly suggest to disallow by default.

**Comment by pedrobaeza** [comment]
But it's already the default behavior in this module AFAIK.

**Comment by levkar** [comment]
Sorry, I get it wrong. It's just for tests. No problem.

**Comment by pedrobaeza** [comment]
@alexis-via would you be able to clean the commits and cherry-pick 73fa871 for having this module ready to be merged, or should we superseed it?

**Review by pedrobaeza — APPROVED** [review]
I have fixed remaining comments (rebase + squash + cherry-pick), so I merge this.

**Pull request merged**